File: docs/js/openalex.js
// OpenAlex client — runs entirely in the browser (OpenAlex is CORS-enabled).
// Mirrors the Python `archive_crawler` OpenAlex source: builds filtered queries,
// reconstructs abstracts from the inverted index, and normalizes papers.

const API_BASE = "https://api.openalex.org";

// Only fetch the fields we actually render — keeps responses small and fast.
const SELECT_FIELDS = [
  "id", "doi", "display_name", "title", "publication_year", "publication_date",
  "type", "authorships", "abstract_inverted_index", "keywords", "concepts",
  "primary_location", "best_oa_location", "open_access", "cited_by_count",
].join(",");

function mailtoParam(mailto) {
  return mailto ? `&mailto=${encodeURIComponent(mailto)}` : "";
}

/** Rebuild a plain-text abstract from OpenAlex's inverted index. */
export function reconstructAbstract(inverted) {
  if (!inverted) return null;
  const positions = [];
  for (const [word, idxs] of Object.entries(inverted)) {
    for (const i of idxs) positions.push([i, word]);
  }
  if (!positions.length) return null;
  positions.sort((a, b) => a[0] - b[0]);
  return positions.map((p) => p[1]).join(" ");
}

function cleanDoi(doi) {
  if (!doi) return null;
  return doi.replace("https://doi.org/", "").trim() || null;
}

/** Normalize a raw OpenAlex "work" into our paper shape. */
export function parseWork(work) {
  const authors = (work.authorships || []).map((a) => {
    const author = a.author || {};
    const inst = (a.institutions || [])[0];
    return {
      name: author.display_name || "Unknown",
      affiliation: inst ? inst.display_name : null,
      orcid: author.orcid || null,
    };
  });

  let keywords = (work.keywords || [])
    .map((k) => k.display_name)
    .filter(Boolean);
  if (!keywords.length) {
    keywords = (work.concepts || [])
      .filter((c) => c.display_name && (c.score || 0) >= 0.3)
      .map((c) => c.display_name);
  }

  const primary = work.primary_location || {};
  const source = primary.source || {};
  const bestOa = work.best_oa_location || {};
  const oaUrl =
    (work.open_access || {}).oa_url ||
    bestOa.pdf_url ||
    bestOa.landing_page_url ||
    null;

  return {
    id: work.id,
    title: work.display_name || work.title || "Untitled",
    authors,
    year: work.publication_year || null,
    date: work.publication_date || null,
    type: work.type || null,
    abstract: reconstructAbstract(work.abstract_inverted_index),
    keywords,
    doi: cleanDoi(work.doi),
    venue: source.display_name || null,
    url: primary.landing_page_url || work.id,
    openAccessUrl: oaUrl,
    isOpenAccess: !!(work.open_access || {}).is_oa,
    citationCount: work.cited_by_count ?? null,
    source: "openalex",
  };
}

const _sourceCache = new Map();

/** Search OpenAlex "sources" (venues/journals) by name. */
export async function searchSources(query, { mailto, perPage = 15 } = {}) {
  if (!query || !query.trim()) return [];
  const url =
    `${API_BASE}/sources?search=${encodeURIComponent(query)}` +
    `&per-page=${perPage}&select=id,display_name,works_count,type,host_organization_name` +
    mailtoParam(mailto);
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`Venue lookup failed (HTTP ${resp.status})`);
  const data = await resp.json();
  return (data.results || []).map((s) => ({
    id: s.id.split("/").pop(),
    name: s.display_name,
    worksCount: s.works_count || 0,
    type: s.type,
    publisher: s.host_organization_name || null,
  }));
}

/**
 * Score how well a candidate source matches a target venue name. Name match
 * dominates (an exact/near-exact title beats a huge unrelated journal), with
 * works-count and conference type as gentle tie-breakers. This fixes chips
 * like "ACM CHI" that previously resolved to whichever match had the most
 * works rather than the actual proceedings.
 */
function scoreSource(cand, query) {
  const n = (cand.name || "").toLowerCase().trim();
  const q = query.toLowerCase().trim();
  let score = 0;
  if (n === q) score += 1000;
  else if (n.includes(q)) score += 500;
  const qTokens = q.split(/[^a-z0-9]+/).filter((t) => t.length > 2);
  if (qTokens.length) {
    const overlap = qTokens.filter((t) => n.includes(t)).length / qTokens.length;
    score += overlap * 300;
  }
  if (cand.type && /conference|proceedings/i.test(cand.type)) score += 40;
  // Log-scaled so a mega-journal can't overpower a strong name match.
  score += Math.log10((cand.worksCount || 0) + 1) * 8;
  return score;
}

/** Resolve a free-text venue name to the best-matching source id. */
export async function resolveVenue(name, opts = {}) {
  if (/^S\d+$/.test(name)) return { id: name, name };
  if (_sourceCache.has(name)) return _sourceCache.get(name);
  const matches = await searchSources(name, { ...opts, perPage: 25 });
  if (!matches.length) throw new Error(`No venue matched "${name}"`);
  const best = matches.reduce((a, b) =>
    scoreSource(b, name) > scoreSource(a, name) ? b : a
  );
  const resolved = { id: best.id, name: best.name };
  _sourceCache.set(name, resolved);
  return resolved;
}

/**
 * Search works. Returns { papers, total, page, perPage, hasMore }.
 * `filters` supports: query, venueIds (array, OR-ed), yearFrom, yearTo,
 * openAccessOnly, sort.
 */
export async function searchWorks(filters = {}, { mailto, page = 1, perPage = 25 } = {}) {
  const {
    query,
    venueIds = [],
    yearFrom,
    yearTo,
    openAccessOnly = false,
    sort = "relevance",
  } = filters;

  const filterParts = [];
  // OpenAlex OR-s values within one filter key using "|".
  if (venueIds.length) {
    filterParts.push(`primary_location.source.id:${venueIds.join("|")}`);
  }
  if (yearFrom) filterParts.push(`from_publication_date:${yearFrom}-01-01`);
  if (yearTo) filterParts.push(`to_publication_date:${yearTo}-12-31`);
  if (openAccessOnly) filterParts.push("open_access.is_oa:true");

  const sortMap = {
    relevance: query ? "relevance_score:desc" : "cited_by_count:desc",
    newest: "publication_date:desc",
    oldest: "publication_date:asc",
    citations: "cited_by_count:desc",
  };

  const params = new URLSearchParams();
  params.set("per-page", String(perPage));
  params.set("page", String(page));
  params.set("select", SELECT_FIELDS);
  params.set("sort", sortMap[sort] || sortMap.relevance);
  if (filterParts.length) params.set("filter", filterParts.join(","));
  if (query && query.trim()) params.set("search", query.trim());

  let url = `${API_BASE}/works?${params.toString()}${mailtoParam(mailto)}`;

  const resp = await fetch(url);
  if (!resp.ok) {
    let detail = "";
    try {
      detail = (await resp.json()).message || "";
    } catch (_) { /* ignore */ }
    throw new Error(`Search failed (HTTP ${resp.status})${detail ? ": " + detail : ""}`);
  }
  const data = await resp.json();
  const total = (data.meta || {}).count || 0;
  const papers = (data.results || []).map(parseWork);
  return {
    papers,
    total,
    page,
    perPage,
    hasMore: page * perPage < total,
  };
}

File: archive_crawler/sources/openalex.py
"""OpenAlex metadata source.

OpenAlex (https://openalex.org) is a free, open catalog of scholarly works.
It is the default backend because it returns every field the crawler targets:
abstract (as an inverted index we reconstruct), keywords/concepts, authors,
publication year, venue, and an open-access URL when one exists. No API key is
required; supplying a ``mailto`` opts into the faster "polite pool".
"""

from __future__ import annotations

import re
from typing import Any, Iterator, Optional

from ..models import Author, Paper
from .base import Source, SourceError

API_BASE = "https://api.openalex.org"


class OpenAlexSource(Source):
    name = "openalex"

    # -- public API --------------------------------------------------------
    def resolve_venue(self, venue: str) -> tuple[Optional[str], str]:
        """Resolve a free-text venue name to an OpenAlex source id.

        Returns ``(source_id, display_name)``. If ``venue`` already looks like
        an OpenAlex source id (e.g. ``S4306420956`` or a full URL), it is used
        verbatim. Otherwise the sources endpoint is searched and the highest
        works-count match is chosen.
        """
        vid = venue.rsplit("/", 1)[-1]
        if re.fullmatch(r"S\d+", vid):
            return vid, venue

        data = self._get_json(
            f"{API_BASE}/sources",
            params=self._with_mailto({"search": venue, "per-page": 25}),
        )
        results = data.get("results") or []
        if not results:
            raise SourceError(f"No OpenAlex venue matched {venue!r}")
        # Prefer the most-published matching source.
        best = max(results, key=lambda s: s.get("works_count", 0))
        return best["id"].rsplit("/", 1)[-1], best.get("display_name", venue)

    def search(
        self,
        query: Optional[str] = None,
        *,
        venue: Optional[str] = None,
        year_from: Optional[int] = None,
        year_to: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> Iterator[Paper]:
        filters: list[str] = []

        if venue:
            source_id, _ = self.resolve_venue(venue)
            filters.append(f"primary_location.source.id:{source_id}")
        if year_from:
            filters.append(f"from_publication_date:{year_from}-01-01")
        if year_to:
            filters.append(f"to_publication_date:{year_to}-12-31")

        params: dict[str, Any] = {
            "per-page": 200,
            "cursor": "*",
        }
        if filters:
            params["filter"] = ",".join(filters)
        if query:
            params["search"] = query

        yielded = 0
        while True:
            page = self._get_json(
                f"{API_BASE}/works", params=self._with_mailto(dict(params))
            )
            results = page.get("results") or []
            if not results:
                break
            for work in results:
                yield self._parse_work(work)
                yielded += 1
                if limit is not None and yielded >= limit:
                    return
            next_cursor = (page.get("meta") or {}).get("next_cursor")
            if not next_cursor:
                break
            params["cursor"] = next_cursor

    # -- helpers -----------------------------------------------------------
    def _with_mailto(self, params: dict[str, Any]) -> dict[str, Any]:
        if self.mailto:
            params.setdefault("mailto", self.mailto)
        return params

    def _parse_work(self, work: dict[str, Any]) -> Paper:
        authors = []
        for a in work.get("authorships") or []:
            author = a.get("author") or {}
            institutions = a.get("institutions") or []
            affiliation = institutions[0]["display_name"] if institutions else None
            authors.append(
                Author(
                    name=author.get("display_name") or "Unknown",
                    affiliation=affiliation,
                    orcid=author.get("orcid"),
                )
            )

        keywords = [
            kw.get("display_name")
            for kw in (work.get("keywords") or [])
            if kw.get("display_name")
        ]
        # Fall back to concepts if the newer keywords field is absent.
        if not keywords:
            keywords = [
                c.get("display_name")
                for c in (work.get("concepts") or [])
                if c.get("display_name") and (c.get("score") or 0) >= 0.3
            ]

        primary = work.get("primary_location") or {}
        source_meta = primary.get("source") or {}
        best_oa = work.get("best_oa_location") or {}
        oa_url = (
            (work.get("open_access") or {}).get("oa_url")
            or best_oa.get("pdf_url")
            or best_oa.get("landing_page_url")
        )

        return Paper(
            title=work.get("display_name") or work.get("title") or "Untitled",
            authors=authors,
            year=work.get("publication_year"),
            abstract=self._reconstruct_abstract(work.get("abstract_inverted_index")),
            keywords=keywords,
            doi=self._clean_doi(work.get("doi")),
            venue=source_meta.get("display_name"),
            url=primary.get("landing_page_url") or work.get("id"),
            open_access_url=oa_url,
            citation_count=work.get("cited_by_count"),
            source=self.name,
            source_id=work.get("id"),
        )

    @staticmethod
    def _clean_doi(doi: Optional[str]) -> Optional[str]:
        if not doi:
            return None
        return doi.replace("https://doi.org/", "").strip() or None

    @staticmethod
    def _reconstruct_abstract(inverted: Optional[dict[str, list[int]]]) -> Optional[str]:
        """Rebuild plain-text abstract from OpenAlex's inverted index."""
        if not inverted:
            return None
        positions: list[tuple[int, str]] = []
        for word, idxs in inverted.items():
            for i in idxs:
                positions.append((i, word))
        if not positions:
            return None
        positions.sort(key=lambda p: p[0])
        return " ".join(word for _, word in positions)
